File: backend/app/models/cart_item.py
"""
CartItem — a dish line inside a cart with quantity.
"""

from sqlalchemy import Column, DateTime, ForeignKey, Integer, UniqueConstraint
from sqlalchemy.orm import relationship
from sqlalchemy.sql import func

from app.database import Base


class CartItem(Base):
    __tablename__ = "cart_items"
    __table_args__ = (
        UniqueConstraint("cart_id", "dish_id", name="uq_cart_items_cart_dish"),
    )

    id = Column(Integer, primary_key=True, index=True)
    cart_id = Column(Integer, ForeignKey("carts.id", ondelete="CASCADE"), nullable=False, index=True)
    dish_id = Column(Integer, ForeignKey("dishes.id", ondelete="CASCADE"), nullable=False, index=True)
    quantity = Column(Integer, nullable=False, default=1)
    created_at = Column(DateTime(timezone=True), server_default=func.now())

    cart = relationship("Cart", back_populates="items")
    dish = relationship("Dish", back_populates="cart_items")

File: frontend/lib/models/cart_item.dart
import 'dish.dart';
import 'json_parse.dart';

/// Line item from `CartItemResponse`.
class CartItem {
  const CartItem({
    required this.id,
    required this.cartId,
    required this.dishId,
    required this.quantity,
    this.createdAt,
    this.dish,
  });

  final int id;
  final int cartId;
  final int dishId;
  final int quantity;
  final DateTime? createdAt;
  final Dish? dish;

  factory CartItem.fromJson(Map<String, dynamic> json) => CartItem(
        id: parseInt(json['id'], field: 'id'),
        cartId: parseInt(json['cart_id'], field: 'cart_id'),
        dishId: parseInt(json['dish_id'], field: 'dish_id'),
        quantity: parseInt(json['quantity'], field: 'quantity'),
        createdAt: parseDateTimeOrNull(json['created_at']),
        dish: json['dish'] is Map<String, dynamic>
            ? Dish.fromJson(json['dish'] as Map<String, dynamic>)
            : null,
      );
}
